sammo.py

```diff
@@ -239,20 +240,67 @@ def onGpsFrame(
         speed: float = -9999.0,
         course: float = -9999.0,
     ) -> None:
-        self.session.lastGpsInfo = (
-            QgsGeometry.fromPointXY(QgsPointXY(longitude, latitude)),
-            (speed, course),
-        )
+        updated = True
         now = datetime.now()
         gpsNow = datetime(now.year, now.month, now.day, h, m, s)
+
+        if self.session.lastGpsInfo[2] == gpsNow and (
+            speed != -9999.0 or course != -9999.0
+        ):
+            # a GPRMC frame is coming after a GPGGA frame with the same
+            # datetime but speed/course are valid
+            self.session.lastGpsInfo[1] = (speed, course)
+        elif self.session.lastGpsInfo[2] != gpsNow:
+            # a newer GPRMC/GPGGA frame is coming
+            self.session.lastGpsInfo = (
+                QgsGeometry.fromPointXY(QgsPointXY(longitude, latitude)),
+                (speed, course),
```

Comment:
I see a problem here. If we update speed / course with GPGGA frame, we will get -9999 values in lastGpsInfo. And if at the same time, user creates a environment record, the feature will have -9999.0 as speed and course value...
But on the other hand, if there is one GPRMC frame each X time (like the serial GPS), does it make sens to keep speed/course values if it a value that is several minutes old ?
@pblottiere what was the frequency of GPRMC frame emission on the serial GPS ?

Reply:
> what was the frequency of GPRMC frame emission on the serial GPS ?

It's totally erratic

Reply:
> And if at the same time, user creates a environment record, the feature will have -9999.0 as speed and course value...

Do we really want to keep speed and course in this case?

Reply:
Actually (with the globalsat gps), the GPGGA frame is always emitted before the GPRMC frame, but each frame is emitted at each second. Despite of this, there is an interval between the time when the GPGGA frame is processed and the time when the GPRMC is processed. And during this time, the user can create an environment record, but it will have no speed/course value.

I integrated your work to dev branch, but I added a datetime information on speed/course attributes. I update speed/course at each GPRMC frame, but I drop these attributes after one minute if no other GPRMC has been emitted since... We are on a boat, it seem's fair to me.